src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs

```diff
@@ -31,6 +31,7 @@ public class ServiceBase : Component
         private bool _commandPropsFrozen;  // set to true once we've use the Can... properties.
         private bool _disposed;
         private bool _initialized;
+        private object _stopLock = new object();
```

Comment:
The lock was a great addition.

Reply:
This fix didn't specifically require it but it helps users of ServiceBase from tripping over the same problem.  Might be we should lock on all state changes, but for now I focused just on stop since that's documented to be bad to call more than once.